File: Data/Examples/CSharpTester/MCTester/GUI/Managers/Manager_MCNames.cs
using MapCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace MCTester.Managers
{
    public static class Manager_MCNames
    {
        private static Dictionary<uint, string> names;

        public static Dictionary<uint, string> Names
        {
            get { return names; }
            set { names = value; }
        }

        static Manager_MCNames()
        {
            names = new Dictionary<uint, string>();
        }

        public static string GetNameById(uint id)
        {
            if (names.ContainsKey(id))
                return names.FirstOrDefault(x => (x.Key == id)).Value;
            return id.ToString();
        }

        public static string GetMcNameById(uint id)
        {
            if (names.ContainsKey(id))
                return names.FirstOrDefault(x => (x.Key == id)).Value;
            return "";
        }

        public static string GetNameById(uint id,string type)
        {
            return "(" + GetNameById(id) + ") " + type;
        }

        //public static string GetIdFromNameByObject(object obj)
        //{
        //    return GetNameByObject(obj,false);
        //}

        public static string GetDefualtName(object obj)
        {
            StringBuilder type = GetType(obj);
            return GetNameById(uint.Parse(obj.GetHashCode().ToString()),type.ToString());
        }

        public static string GetNameByObject(object obj)
        {
            StringBuilder objType = GetType(obj);
            return GetNameByObject(obj, objType.ToString());
        }

        public static void GetNameByObjectArray(object obj, string[] arrOutput)
        {
            StringBuilder objType = GetType(obj);
            string id = GetIdByObject(obj);
            arrOutput[0] = id;
            arrOutput[1] = id + " " + objType;
        }

        public static void GetNameByObjectArrayByNameAndId(object obj, string[] arrOutput)
        {
            StringBuilder objType = GetType(obj);
            string id = GetIdByObject(obj);
            arrOutput[0] = id;
            arrOutput[1] = objType.ToString();
        }

        public static string GetIdByObject(object obj)
        {
            uint u_id = (uint)obj.GetHashCode();

            string fullName; 
            string mcName = string.Empty;

            if (obj is IDNMcObjectScheme)
            {
                try
                {
                    mcName = ((IDNMcObjectScheme)obj).GetName();
                }
                catch (MapCoreException McEx)
                {
                    MapCore.Common.Utilities.ShowErrorMessage("IDNMcObjectScheme.GetName()", McEx);
                }
            }
            else if (obj is IDNMcObjectSchemeNode)
            {
                try
                {
                    mcName = ((IDNMcObjectSchemeNode)obj).GetName();
                }
                catch (MapCoreException McEx)
                {
                    MapCore.Common.Utilities.ShowErrorMessage("IDNMcObjectSchemeNode.GetName()", McEx);
                }
            }
            else if (obj is IDNMcConditionalSelector)
            {
                try
                {
                    mcName = ((IDNMcConditionalSelector)obj).GetName();
                }
                catch (MapCoreException McEx)
                {
                    MapCore.Common.Utilities.ShowErrorMessage("IDNMcConditionalSelector.GetName()", McEx);
                }
            }
            else if (obj is IDNMcObject)
            {
                try
                {
                    string name, desc;
                    ((IDNMcObject)obj).GetNameAndDescription(out name, out desc);
                    mcName = name;
                }
                catch (MapCoreException McEx)
                {
                    MapCore.Common.Utilities.ShowErrorMessage("IDNMcObject.GetNameAndDescription()", McEx);
                }
            }

            if (mcName != string.Empty)
                fullName = "\"" +mcName+ "\"";
            else if (names.ContainsKey(u_id))
                fullName = "\"" + names.FirstOrDefault(x => (x.Key == u_id)).Value + "\"";
            else
                fullName = u_id.ToString();

            
            return fullName;
        }


        public static string GetNameByObject(object obj, string objType)
        {
            string fullName = "(" + GetIdByObject(obj) + ")";
            string strMoreData = "";
            if (obj is IDNMcObject)
            {
                try
                {
                    DNESymbologyStandard symbologyStandard = ((IDNMcObject)obj).GetSymbologyStandard();
                    if (symbologyStandard != DNESymbologyStandard._ESS_NONE)
                    {
                        strMoreData = " [" + symbologyStandard.ToString().Replace("_ESS_", "") + "]";
                    }
                }
                catch (MapCoreException McEx)
                {
                    MapCore.Common.Utilities.ShowErrorMessage("GetSymbologyStandard", McEx);
                }
            }
            else if(obj is IDNMcMapViewport)
            {
                IDNMcMapViewport mapViewport = obj as IDNMcMapViewport;
                DNEMapType mapType = mapViewport.MapType; 
                string strImage = (mapViewport.GetImageCalc() != null) ? " Image" : "";
                strMoreData = " [" + mapType.ToString().Replace("_EMT_", "") + strImage + "]";
            }
            else if (obj is IDNMcMapGrid)
            {
                IDNMcMapGrid mapGrid = obj as IDNMcMapGrid;
                DNSGridRegion[] gridRegions = mapGrid.GetGridRegions();
                if (gridRegions != null && gridRegions.Length == 1 && gridRegions[0].pCoordinateSystem != null)
                    strMoreData = " [" + gridRegions[0].pCoordinateSystem.GetGridCoorSysType().ToString().Replace("_EGCS_", "")+ "]";
                
            }
            return fullName + " " + objType + strMoreData;
        }
        
        public static StringBuilder GetType(object obj)
        {
            string TypeName = GeneralFuncs.GetDirectInterfaceName(obj.GetType());
            TypeName = TypeName.Replace("IDNMc", "");
            StringBuilder modifiedString = new StringBuilder();
            for(int i=0;i<TypeName.Length;i++)
            {
                char s = TypeName[i];
                if (char.IsDigit(s) && s == '3' && TypeName.ToLower().Contains("3d"))
                {
                    modifiedString.Append(" 3D");
                    i++;
                }
                else
                {
                    if (char.IsUpper(s))
                        modifiedString.Append(' ');
                    modifiedString.Append(s);
                }
            }
           
            return modifiedString;
        }

        public static void SetName(object node, string newName)
        {
            if (!SetMcName(node, newName))
            {
                uint id = uint.Parse(node.GetHashCode().ToString()); 
                
                if (names.ContainsKey(id))
                    names[id] = newName;
                else
                    names.Add(id, newName);
            }
        }

        private static bool SetMcName(object node, string newName)
        {
            bool isChanged = false;
            if (node is IDNMcObjectScheme)
            {
                try
                {
                    isChanged = true;
                    ((IDNMcObjectScheme)node).SetName(newName);
                }
                catch (MapCoreException McEx)
                {
                    MapCore.Common.Utilities.ShowErrorMessage("IDNMcObjectScheme.SetName()", McEx);
                }
            }
            if (node is IDNMcObjectSchemeNode)
            {
                try
                {
                    isChanged = true;
                    ((IDNMcObjectSchemeNode)node).SetName(newName);
                }
                catch (MapCoreException McEx)
                {
                    MapCore.Common.Utilities.ShowErrorMessage("IDNMcObjectSchemeNode.SetName()", McEx);
                }
            }
            if (node is IDNMcConditionalSelector)
            {
                try
                {
                    isChanged = true;
                    ((IDNMcConditionalSelector)node).SetName(newName);
                }
                catch (MapCoreException McEx)
                {
                    MapCore.Common.Utilities.ShowErrorMessage("IDNMcConditionalSelector.SetName()", McEx);
                }
            }
            if(node is IDNMcObject)
            {
                try
                {
                    isChanged = true;
                    string name, desc;
                    ((IDNMcObject)node).GetNameAndDescription(out name, out desc);

                    ((IDNMcObject)node).SetNameAndDescription(newName, desc);
                }
                catch (MapCoreException McEx)
                {
                    MapCore.Common.Utilities.ShowErrorMessage("IDNMcObject.Get/SetNameAndDescription()", McEx);
                }
                
            }
            return isChanged;
        }

        public static bool RemoveName(object node)
        {
            uint id = uint.Parse(node.GetHashCode().ToString());
            bool isRemoved = false;
            if (SetMcName(node, string.Empty))
                isRemoved = true;
            if (!isRemoved && names.ContainsKey(id))
            {
                names.Remove(id);
                isRemoved = true;
            }
            return isRemoved;
        }

    }
}

File: Data/Examples/Android/McAndroidTester/app/src/main/java/com/elbit/mapcore/mcandroidtester/managers/Manager_MCNames.java
package com.elbit.mapcore.mcandroidtester.managers;

import com.elbit.mapcore.mcandroidtester.utils.AlertMessages;
import com.elbit.mapcore.mcandroidtester.utils.BaseApplication;
import com.elbit.mapcore.mcandroidtester.utils.Consts;

import java.util.Hashtable;

import com.elbit.mapcore.General.MapCoreException;
import com.elbit.mapcore.Interfaces.OverlayManager.IMcConditionalSelector;
import com.elbit.mapcore.Interfaces.OverlayManager.IMcObjectLocation;
import com.elbit.mapcore.Interfaces.OverlayManager.IMcObjectScheme;
import com.elbit.mapcore.Interfaces.OverlayManager.IMcObjectSchemeNode;
import com.elbit.mapcore.mcandroidtester.utils.Funcs;

/**
 * Created by tc99382 on 24/01/2017.
 */
public class Manager_MCNames {
    private static Hashtable<Integer, String> names;
    private static Manager_MCNames instance;

    private Manager_MCNames() {
        names = new Hashtable<>();
    }

    public static Hashtable<Integer, String> getNames() {
        return names;
    }

    public static void setNames(Hashtable<Integer, String> names) {
        Manager_MCNames.names = names;
    }

    public String getNameById(Integer id) {
        if (names.containsKey(id))
            return names.get(id);
        return String.valueOf(id);
    }

    public String getNameById(Integer id, String type) {
        return "(" + getNameById(id) + ") " + type;
    }

    public String getNameByObject(Object obj) {
        StringBuilder objType = getType(obj);
        return getNameByObject(obj, objType.toString());
    }

    public String[] getNameByObjectArray(Object obj, String[] arrOutput) {
        StringBuilder objType = getType(obj);
        String id = getIdByObject(obj);
        arrOutput[0] = id;
        arrOutput[1] = id + " " + objType;
        return arrOutput;
    }


    private StringBuilder getType(Object obj) {
        String typeName = obj.getClass().getSimpleName();
        typeName = typeName.replace("Mc", "");
        StringBuilder modifiedString = new StringBuilder();
        for (Character s : typeName.toCharArray()) {
            if (Character.isUpperCase(s))
                modifiedString.append(' ');
            modifiedString.append(s);
        }
        return modifiedString;
    }

    public String getNameByObject(Object obj, String objType) {
        String fullName = "(" + getIdByObject(obj) + ")";
        return fullName + " " + objType;
    }

    public String getIdByObject(Object obj) {
        Integer id = obj.hashCode();
        String fullName;
        String mcName = Consts.EMPTY_STRING;
        if (obj instanceof IMcObjectScheme) {
            try {
                mcName = (((IMcObjectScheme) obj).GetName());
            } catch (MapCoreException e) {
                AlertMessages.ShowMapCoreErrorMessage(BaseApplication.getCurrActivityContext(), e, "IMcObjectScheme.GetName()");
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        } else if (obj instanceof IMcObjectLocation) {
            try {
                mcName = ((IMcObjectLocation) obj).GetName();
            } catch (MapCoreException mcEx) {
                AlertMessages.ShowMapCoreErrorMessage(BaseApplication.getCurrActivityContext(), mcEx, "IMcObjectLocation.GetName()");
            } catch (Exception e) {
                e.printStackTrace();
            }
        } else if (obj instanceof IMcObjectSchemeNode) {
            try {
                mcName = ((IMcObjectSchemeNode) obj).GetName();
            } catch (MapCoreException mcEx) {
                AlertMessages.ShowMapCoreErrorMessage(BaseApplication.getCurrActivityContext(), mcEx, "IMcObjectSchemeNode.GetName()");
            } catch (Exception e) {
                e.printStackTrace();
            }
        } else if (obj instanceof IMcConditionalSelector) {

            try {
                mcName = ((IMcConditionalSelector) obj).GetName();
            } catch (MapCoreException mcEx) {
                AlertMessages.ShowMapCoreErrorMessage(BaseApplication.getCurrActivityContext(), mcEx, "IMcConditionalSelector.GetName()");
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        if (mcName != Consts.EMPTY_STRING && mcName != null) {
            fullName = mcName;
        } else if (names.containsKey(id)) {
            fullName = names.get(id);
        } else
            fullName = String.valueOf(id);
        return fullName;

    }

    private String getDefaultName(Object obj) {
        StringBuilder objType = getType(obj);
        return getNameByObject(obj, objType.toString());
    }

    public void setName(Object node, String newName) {

        if (!setMcName(node, newName)) {
            Integer id = node.hashCode();
            names.put(id, newName);
        }
    }

    private boolean setMcName(Object node, String newName) {
         boolean isChanged = false;
        if (node instanceof IMcObjectScheme) {
            isChanged = true;
            try {

                ((IMcObjectScheme) node).SetName(newName);
            } catch (MapCoreException McEx) {
                AlertMessages.ShowMapCoreErrorMessage(BaseApplication.getCurrActivityContext(), McEx, "IDNMcObjectScheme.SetName()");
            } catch (Exception e) {

            }
        }
        if (node instanceof IMcObjectSchemeNode) {
            isChanged = true;
            try {
                ((IMcObjectSchemeNode) node).SetName(newName);
            } catch (MapCoreException McEx) {
                AlertMessages.ShowMapCoreErrorMessage(BaseApplication.getCurrActivityContext(), McEx, "IDNMcObjectSchemeNode.SetName()");
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        if (node instanceof IMcConditionalSelector) {
            isChanged = true;
            try {
                ((IMcConditionalSelector) node).SetName(newName);
            } catch (MapCoreException McEx) {
                AlertMessages.ShowMapCoreErrorMessage(BaseApplication.getCurrActivityContext(), McEx, "IDNMcConditionalSelector.SetName()");
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        return isChanged;
    }

    public boolean removeName(Object node) {
        Integer id = Integer.valueOf(node.hashCode());
        boolean isRemoved = false;
        if (setMcName(node, Consts.EMPTY_STRING))
            isRemoved = true;
        if (!isRemoved && names.containsKey(id)) {
            names.remove(id);
            isRemoved = true;
        }
        return isRemoved;
    }

    public static Manager_MCNames getInstance() {
        if (instance == null)
            instance = new Manager_MCNames();
        return instance;
    }


}
